src/Domain/Services/DraftOrders.php

```diff
@@ -0,0 +1,208 @@
+		}
+	}
+
+	/**
+	 * Unschedule cron jobs that are present.
+	 */
+	protected function maybe_remove_cronjobs() {
+		if ( function_exists( 'as_next_scheduled_action' ) && true === as_next_scheduled_action( 'woocommerce_cleanup_draft_orders' ) ) {
+			as_unschedule_all_actions( 'woocommerce_cleanup_draft_orders' );
+		}
+	}
+
+	/**
+	 * Register custom order status for orders created via the API during checkout.
+	 *
+	 * Draft order status is used before payment is attempted, during checkout, when a cart is converted to an order.
+	 *
+	 * @param array $statuses Array of statuses.
+	 * @internal
+	 * @return array
+	 */
+	public function register_draft_order_status( array $statuses ) {
+		$statuses[ self::DB_STATUS ] = _x( 'Draft', 'Order status', 'woo-gutenberg-products-block' );
+		return $statuses;
+	}
+
+	/**
+	 * Register custom order post status for orders created via the API during checkout.
+	 *
+	 * @param array $statuses Array of statuses.
+	 * @internal
+
+	 * @return array
+	 */
+	public function register_draft_order_post_status( array $statuses ) {
+		$statuses[ self::DB_STATUS ] = [
+			'label'                     => _x( 'Draft', 'Order status', 'woo-gutenberg-products-block' ),
+			'public'                    => false,
+			'exclude_from_search'       => false,
+			'show_in_admin_all_list'    => false,
+			'show_in_admin_status_list' => true,
+			/* translators: %s: number of orders */
+			'label_count'               => _n_noop( 'Drafts <span class="count">(%s)</span>', 'Drafts <span class="count">(%s)</span>', 'woo-gutenberg-products-block' ),
+		];
+		return $statuses;
+	}
+
+	/**
+	 * Append draft status to a list of statuses.
+	 *
+	 * @param array $statuses Array of statuses.
+	 * @internal
+
+	 * @return array
+	 */
+	public function append_draft_order_post_status( $statuses ) {
+		$statuses[] = self::STATUS;
+		return $statuses;
+	}
+
+	/**
+	 * Delete draft orders older than a day in batches of 20.
+	 *
+	 * Ran on a daily cron schedule.
+	 *
+	 * @internal
+	 */
+	public function delete_expired_draft_orders() {
+		$count      = 0;
+		$batch_size = 20;
+		$orders     = wc_get_orders(
+			[
+				'date_modified' => '<=' . strtotime( '-1 DAY' ),
+				'limit'         => $batch_size,
+				'status'        => self::DB_STATUS,
+				'type'          => 'shop_order',
+			]
+		);
+
+		// do we bail because the query results are unexpected?
+		try {
+			$this->assert_order_results( $orders, $batch_size );
+			if ( $orders ) {
+				foreach ( $orders as $order ) {
+					$order->delete( true );
+					$count ++;
+				}
+			}
+			if ( $batch_size === $count && function_exists( 'as_enqueue_async_action' ) ) {
+				as_enqueue_async_action( 'woocommerce_cleanup_draft_orders' );
+			}
+		} catch ( Exception $error ) {
+			// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+			error_log( $error->getMessage() );
```

Comment:
This should be wrapped with a check IMO `if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {`

Reply:
Or we could use https://github.com/woocommerce/woocommerce/blob/dd49f89e77acaaa097251fe0a5dd69320ded48c6/includes/wc-deprecated-functions.php#L92

Reply:
I normally agree on logging only on non-production, but this would be triggered in an exceptional case where the logs can serve as a troubleshooting tool for why potentially no draft orders were cleaned up on this schedule.